### ISSUE TITLE
Add a message to a web socket

### ISSUE THREAD

**Issue by ramvasanth** [eranyanay/1m-go-websockets#11]
Hi , we maintain millions of web socket connection without using epoll. if i adopt your proposed solution, i have one question.

how do i get a specific web socket from the epoll and write a message into it.  please advice me,

**Comment by eranyanay** [comment]
@ramvasanth 
Using epoll means that when it returns, you know a specific fd is ready to read
This means that you need to have some mapping from fd to the connection, such as `conns map[int]net.Conn` (type probably depends on the chosen ws library)

So you will most probably do something similar to [this](https://github.com/eranyanay/1m-go-websockets/blob/master/3_optimize_ws_goroutines/epoll.go#L60)

Makes sense?

**Comment by jerber123** [comment]
**follow the instructions step by step using the server "4_optimize_gobwas"**
1.- import "strings", like this:
```
import ( 
	"strings"
)
```

2 .- Right after the "epoller" variable create 3 variables, like this:
```
var epoller *epoll
var x = map[string]int{}
var xx = map[int]net.Conn{}
var xxx = map[int]string{}
```

3.- Inside the wsHandler function before calling "epoller.Add" create a "token" variable, like this:
```
token := r.FormValue("token") 
if err := epoller.Add(conn,token); err != nil {
```

4.- After calling "epoller.Add" add the following code, like this:
```
	token := r.FormValue("token") 
	if err := epoller.Add(conn,token); err != nil {
		log.Printf("Failed to add connection %v", err)
		conn.Close()
	} 
	go func() {
		defer conn.Close()

		for {
			msg, op, err := wsutil.ReadClientData(conn)
			if err != nil {
				// handle error
			}
			msgstring := string(msg)
			ss := strings.Split(msgstring, "|")  
			if ss[0] != "" {
				if xx[x[ss[1]]] != nil {
				   	err = wsutil.WriteServerMessage(xx[x[ss[1]]], op, msg)
					if err != nil {
						// handle error
					} 
				}
			} 

 
		}
	}()
```

5.- Add "token string" where the "Add" function is defined, like this:
`func (e *epoll) Add(conn net.Conn, token string ) error {`

6.- Within the "Add" function, at the end but before "return nil" define the three variables, like this:
```
x[token] = fd
xx[fd] = conn
xxx[fd] = token
return nil
```

7.- Inside the "Remove" function before "delete(e.connections, fd)" place the following code, like this:
```
delete(x,xxx[fd])
delete(xx,fd)
delete(xxx,fd)
delete(e.connections, fd)
```

Finally on client 1 establish the connection, like this:
`var conn = new WebSocket("ws://18.223.247.37:5002/?token=mytoken1"); `

Send message like this:
`conn.send("mytoken1|mytoken2|mensage here");` 

On client 2 establish the connection, like this:
`var conn = new WebSocket("ws://18.223.247.37:5002/?token=mytoken2"); `

Send message like this:
`conn.send("mytoken2|mytoken1|mensage here");` 

In summary, 3 maps are created although it would weigh a lot with many connections.
Can anyone tell me if what I did is correct?

**Comment by eranyanay** [comment]
@jerber123 not sure what you're trying to solve
specifically, the epoll instance demonstrated in this repo already maps a given websocket fd to its connection interface.
If you need additional data, you can wrap the map value from `net.Conn` to some custom struct of your choice.
There is no need for three different maps, only one, and it involves an allocation of an integer and a struct of your choice, which shouldnt be more than a few pointers at best

**Comment by jerber123** [comment]
I added a map (token) to epoll with the value "fd"
```
type epoll struct {
	fd          int
	connections map[int]net.Conn 
	token		map[string]int 
	tokeninv	map[int]string
	lock        *sync.RWMutex
}
```

but to remove it when closing the connection I need to create another map
(token) inverted,
https://github.com/jerber123/1m-ws-con/blob/master/server.go#L175
Is this the best way?
Could you give an example of your recommendation for a better way to do it?
demo: https://github.com/jerber123/1m-ws-con

**Comment by eranyanay** [comment]
What id do you need?
Use the fd, it's unique
